### ISSUE TITLE
Zoom does not return the same values when resizing a sub-array relative to the entire array

### ISSUE THREAD

**Issue by GeertLitjens** [scipy/scipy#5223]
When you resize a sub-array with a factor of 3 (or higher) the result should be the same when resizing the entire array and then taking the same sub-array (3 times larger of course). However, when using zoom it is not, regardless of interpolation order

Code to reproduce:

```
import numpy
from scipy.ndimage import zoom
from scipy.misc import imresize
arr = numpy.array(range(36))
arr.shape = (6,6)
res = zoom(arr, 3., order=0)
res_sub = zoom(arr[0:4,0:4], 3., order=0)
numpy.abs(res[0:12,0:12] - res_sub).sum()
```

The last result should return zero, but it does not. When using imresize, the result is as expected:

```
arr = numpy.array(range(36))
arr.shape = (6,6)
res = imresize(arr, 3., interp='nearest', mode='F').astype(int)
res_sub = imresize(arr[0:4,0:4], 3., interp='nearest', mode='F').astype(int)
numpy.abs(res[0:12,0:12] - res_sub).sum()
```

**Comment by GeertLitjens** [comment]
Since there is no response, I'm wondering whether I reported this bug correctly. Can others not replicate it? Or is zoom not a high-priority? If so, are there other alternatives than imresize?

**Comment by argriffing** [comment]
> Since there is no response, I'm wondering whether I reported this bug correctly.

Yes this is reported correctly. The maintainers listed for this module are @stefanv and @thouis.  Maybe they are on vacation or are busy with https://github.com/scikit-image/scikit-image?

**Comment by stefanv** [comment]
This won't happen with, e.g., ``skimage.transform.rescale`` either.  My first thought was that this is because ndimage thinks of data values as the centres of pixels.  But I've now simulated several indexing mechanisms and none of them seem right:

```
from scipy import ndimage as ndi
import numpy as np

y = np.array([1, 2, 3, 4])

print('Method: ndimage')
print(ndi.zoom(y, 3, order=0))

print('\n---------------------------------------------------')
print('Method: center coordinates')
print('Image looks like this:')
print('''
Pixels:         [       ][       ][       ][       ]
Indices:            0        1        2        3
Zoomed pixels:  [ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ]
Zoomed indices:  0  1  2  3  4  5  6 ...
''')
delta = (4 / 12) / 2
x = np.linspace(delta, 3.5 - delta, 4 * 3, endpoint=True)
print(y[(x).astype(int)])

print('\n---------------------------------------------------')
print('Method: data coordinates')
print('Image looks like this:')
print('''
Pixels:          .      .      .       .
Indices:         0      1      2       3
Zoomed pixels:   . . . . . . . . . . . .
Zoomed indices:  0 1 2 3 4 5 6 ...
''')
x = np.linspace(0, 3, 4 * 3, endpoint=True)
print(y[x.astype(int)])

print('\n---------------------------------------------------')
print('Method: geometric data coordinates')
print('Image looks like this:')
print('''
Pixels:          .     .     .     .
Indices:         0     1     2     3
Zoomed pixels:   . . . . . . . . . . . .
Zoomed indices:  0 1 2 3 4 5 6 ...
''')
x = np.linspace(0, 11/3, 4 * 3, endpoint=True)
print(y[x.astype(int)])
```

Output:

```
Method: ndimage
[1 1 2 2 2 2 3 3 3 3 4 4]

---------------------------------------------------
Method: center coordinates
Image looks like this:

Pixels:         [       ][       ][       ][       ]
Indices:            0        1        2        3
Zoomed pixels:  [ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ]
Zoomed indices:  0  1  2  3  4  5  6 ...

[1 1 1 2 2 2 2 3 3 3 4 4]

---------------------------------------------------
Method: data coordinates
Image looks like this:

Pixels:          .      .      .       .
Indices:         0      1      2       3
Zoomed pixels:   . . . . . . . . . . . .
Zoomed indices:  0 1 2 3 4 5 6 ...

[1 1 1 1 2 2 2 2 3 3 3 4]

---------------------------------------------------
Method: geometric data coordinates
Image looks like this:

Pixels:          .     .     .     .
Indices:         0     1     2     3
Zoomed pixels:   . . . . . . . . . . . .
Zoomed indices:  0 1 2 3 4 5 6 ...

[1 1 1 2 2 2 3 3 3 4 4 4]
```

**Comment by stefanv** [comment]
Of course, I should be rounding, not casting.  That explains it:

```
from scipy import ndimage as ndi
import numpy as np

y = np.array([1, 2, 3, 4])

print('Method: ndimage')
print(ndi.zoom(y, 3, order=0))

print('\n---------------------------------------------------')
print('Method: center coordinates')
print('Image looks like this:')
print('''
Pixels:         [       ][       ][       ][       ]
Indices:            0        1        2        3
Zoomed pixels:  [ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ]
Zoomed indices:  0  1  2  3  4  5  6 ...
''')
delta = (4 / 12) / 2
x = np.linspace(delta, 3.5 - delta, 4 * 3, endpoint=True)
print(y[np.round(x).astype(int)])

print('\n---------------------------------------------------')
print('Method: data coordinates')
print('Image looks like this:')
print('''
Pixels:          .      .      .       .
Indices:         0      1      2       3
Zoomed pixels:   . . . . . . . . . . . .
Zoomed indices:  0 1 2 3 4 5 6 ...
''')
x = np.linspace(0, 3, 4 * 3, endpoint=True)
print(y[np.round(x).astype(int)])

print('\n---------------------------------------------------')
print('Method: geometric data coordinates')
print('Image looks like this:')
print('''
Pixels:          .     .     .     .
Indices:         0     1     2     3
Zoomed pixels:   . . . . . . . . . . . .
Zoomed indices:  0 1 2 3 4 5 6 ...
''')
x = np.linspace(0, 11/3, 4 * 3, endpoint=True)
ix = np.round(x).astype(int)
ix[ix > 3] = 3
print(y[ix])
```

```
Method: ndimage
[1 1 2 2 2 2 3 3 3 3 4 4]

---------------------------------------------------
Method: center coordinates
Image looks like this:

Pixels:         [       ][       ][       ][       ]
Indices:            0        1        2        3
Zoomed pixels:  [ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ]
Zoomed indices:  0  1  2  3  4  5  6 ...

[1 1 2 2 2 3 3 3 3 4 4 4]

---------------------------------------------------
Method: data coordinates
Image looks like this:

Pixels:          .      .      .       .
Indices:         0      1      2       3
Zoomed pixels:   . . . . . . . . . . . .
Zoomed indices:  0 1 2 3 4 5 6 ...

[1 1 2 2 2 2 3 3 3 3 4 4]

---------------------------------------------------
Method: geometric data coordinates
Image looks like this:

Pixels:          .     .     .     .
Indices:         0     1     2     3
Zoomed pixels:   . . . . . . . . . . . .
Zoomed indices:  0 1 2 3 4 5 6 ...

[1 1 2 2 2 3 3 3 4 4 4 4]

```

**Comment by stefanv** [comment]
With `round` defined as:

```
def round(x):
    return (x + 0.5).astype(int)
```

```
Method: ndimage
[1 1 2 2 2 2 3 3 3 3 4 4]

---------------------------------------------------
Method: center coordinates
Image looks like this:

Pixels:         [       ][       ][       ][       ]
Indices:            0        1        2        3
Zoomed pixels:  [ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ][ ]
Zoomed indices:  0  1  2  3  4  5  6 ...

[1 1 2 2 2 3 3 3 3 4 4 4]

---------------------------------------------------
Method: data coordinates
Image looks like this:

Pixels:          .      .      .       .
Indices:         0      1      2       3
Zoomed pixels:   . . . . . . . . . . . .
Zoomed indices:  0 1 2 3 4 5 6 ...

[1 1 2 2 2 2 3 3 3 3 4 4]

---------------------------------------------------
Method: geometric data coordinates
Image looks like this:

Pixels:          .     .     .     .
Indices:         0     1     2     3
Zoomed pixels:   . . . . . . . . . . . .
Zoomed indices:  0 1 2 3 4 5 6 ...

[1 1 2 2 2 3 3 3 4 4 4 4]
```

**Comment by grlee77** [comment]
@rgommers & @stefanv

I took a look at this issue which was linked in #12773 and can confirm that scikit-image's rescale does work as expected here:

```Python
import numpy
from skimage.transform import rescale
arr = numpy.array(range(36))
arr.shape = (6,6)
res = rescale(arr, 3., order=0)
res_sub = rescale(arr[0:4,0:4], 3., order=0)
numpy.abs(res[0:12,0:12] - res_sub).sum()  # should return 0.0
```

Should we change `scipy.ndimage.zoom` to behave in the same way? It is easy to change the function, requiring only a few lines of code to change, but doing so does lead to 37 test failures, so various tests would have to be rewritten and it may also end up breaking some downstream libraries.

For future reference the required changes are:
1.) change the following two lines https://github.com/scipy/scipy/blob/232092afebf7a59949dca5595aaac3446814293c/scipy/ndimage/src/ni_interpolation.c#L784-L785
to
```C
            if (zooms) {
                cc += 0.5;
                cc *= zoom;
                cc -= 0.5;
            }
```

2.) Do not subtract -1 from the shapes in the following two lines:
https://github.com/scipy/scipy/blob/232092afebf7a59949dca5595aaac3446814293c/scipy/ndimage/interpolation.py#L763-L766

**Comment by rgommers** [comment]
> Should we change `scipy.ndimage.zoom` to behave in the same way? It is easy to change the function, requiring only a few lines of code to change, but doing so does lead to 37 test failures, so various tests would have to be rewritten and it may also end up breaking some downstream libraries.

That seems painful. How about just deprecating `zoom` and adding a new `rescale` function with the desired behaviour? There's [a ton of other zoom issues](https://github.com/scipy/scipy/issues?q=is%3Aopen+is%3Aissue+label%3Ascipy.ndimage+zoom), so if we can verify that those don't apply to the new `rescale` function, then I think that would be very much justified.

**Comment by grlee77** [comment]
> That seems painful. How about just deprecating zoom and adding a new rescale function with the desired behaviour? There's a ton of other zoom issues, so if we can verify that those don't apply to the new rescale function, then I think that would be very much justified.

I implemented the suggested change via a `kwarg` in #13095. I left the default behavior the same so nothing will break downstream and the existing test cases still pass. A couple of the new test cases are segfaulting on AArch64, though, and I haven't had a chance to track down why.
